### PR TITLE
Adding support for Symbolic Shapes in Inplace Ops #65642

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65729

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

Differential Revision: [D31961857](https://our.internmc.facebook.com/intern/diff/D31961857)

### CODE REVIEW

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 1]
```diff
@@ -58,6 +58,22 @@ const std::string shape_compute_functions =
         def unary(self: List[int]):
           return _copy(self)
 
+        def broadcast_inplace(a: List[int], b: List[int]):
```
Comment (eellison) [inline torch/csrc/jit/runtime/symbolic_shape_registry.cpp:61]: just curious - did you write this or find logic for it someplace ?

File: torch/csrc/distributed/c10d/ProcessGroup.hpp [hunk 2]
```diff
@@ -226,7 +226,8 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
   // function.
   virtual c10::intrusive_ptr<ProcessGroup::Work> allreduce_coalesced(
       std::vector<at::Tensor>& /* tensors */,
-      const AllreduceCoalescedOptions& /* opts */ = AllreduceCoalescedOptions()) {
```
Comment (eellison) [inline torch/csrc/distributed/c10d/ProcessGroup.hpp:229]: nit unrelated change

File: test/jit/test_symbolic_shape_analysis.py [hunk 3]
```diff
@@ -141,6 +149,31 @@ def apply(fn):
...
+
+        binary_ops = [
+            div_inplace_tensor,
+            add_inplace_tensor,
+        ]
+
+        for fn in binary_ops:
+            size_1 = [4, 4, 8]  # x (can't broadcast because it's an inplace op)
+            size_2 = [4, 1, 8]
+            t = torch.jit.trace(fn, (torch.rand([4]), torch.rand([4])))
```
Comment (eellison) [inline test/jit/test_symbolic_shape_analysis.py:171]: Im pretty sure trace will replace things with their functional equivalent (or it sometimes does this). Can you script this instead ?

File: test/jit/test_symbolic_shape_analysis.py [hunk 4]
```diff
@@ -141,6 +149,31 @@ def apply(fn):
...
+            inputs = list(t.graph.inputs())
+            inputs[0].setType(inputs[0].type().with_sizes(size_1))
+            inputs[1].setType(inputs[1].type().with_sizes(size_2))
```
Comment (eellison) [inline test/jit/test_symbolic_shape_analysis.py:174]: maybe leave the second input untyped to show that we are correctly using the type from the first input ?

**Review by eellison — COMMENTED** [review]
Nice!! looks good - just one question about using FunctionType::isSubtypeOf and it should b ready to land

File: torch/csrc/distributed/c10d/ProcessGroup.hpp (hunk 2 above)
Comment (Gamrix) [reply torch/csrc/distributed/c10d/ProcessGroup.hpp:229]: :(

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp (hunk 1 above)
Comment (Gamrix) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:61]: I just wrote this based on just checking what broadcasts are valid for inplace. 

However, it seems based on the comment you left on test_symbolic:174 that I should not take the second argument into account at all, and instead treat all inplace ops as unary.

File: test/jit/test_symbolic_shape_analysis.py (hunk 4 above)
Comment (Gamrix) [reply test/jit/test_symbolic_shape_analysis.py:174]: as with the other comment, should we treat all inplace ops the same as unary? (if they aren't then they should error out anyways).

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 5]
```diff
@@ -560,7 +576,6 @@ static const OperatorMap<std::string>& get_schema_to_function_graph() {
       {"aten::gt.Tensor(Tensor self, Tensor other) -> Tensor", "broadcast"},
       {"aten::rsub.Tensor(Tensor self, Scalar other, Scalar alpha=1) -> Tensor", "unary"},
       {"aten::add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor", "broadcast"},
-      {"aten::add_.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor", "broadcast"},
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:563]: (The inplace version of them, of course)

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 6]
```diff
@@ -629,6 +644,91 @@ std::unordered_map<const FunctionSchema*, std::shared_ptr<Graph>>
 // CompilationUnit that holds all these Functions and keeps them alive.
 auto compilation_unit = std::make_shared<CompilationUnit>();
 
+const at::optional<const FunctionSchema*> getInplaceVariant(
+    const FunctionSchema& base_schema) {
+  auto& inplace_variants =
+      getAllOperatorsFor(c10::Symbol::fromQualString(base_schema.name() + "_"));
+
+  for (const auto& variant : inplace_variants) {
+    // Need to check that all args are the same except for the first, which
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:653]: Yeah, that would strip out most of the logic.

[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:653]: Ugh, it's tough. @makslevental how did you find this function lol ? (he used it in a different PR)

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:61]: No, this seems right, we want to preserve the error checking

File: test/jit/test_symbolic_shape_analysis.py (hunk 4 above)
Comment (eellison) [reply test/jit/test_symbolic_shape_analysis.py:174]: No, the error checking is good, this would just be related to checking that we are doing right propagation. When I land code for using the partial-evaluated grpah, and I can add a check that we're throwing correclty.

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:651]: I think whenever you're rewriting what seems like it should exist elsewhere, you should ping a chat. this is a case - is this schema equivalent with another schema - which likely already exists.

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 7]
```diff
@@ -629,6 +642,85 @@ std::unordered_map<const FunctionSchema*, std::shared_ptr<Graph>>
...
+    // is almost the same except for the Alias info
+    const FunctionSchema* schema = &variant->schema();
+    // TODO: try this out. If the tests fail, then we know something is wrong
```
Comment (eellison) [inline torch/csrc/jit/runtime/symbolic_shape_registry.cpp:654]: can you remove this comment ?

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 8]
```diff
@@ -629,6 +642,85 @@ std::unordered_map<const FunctionSchema*, std::shared_ptr<Graph>>
...
+    if (!schema->isSubtypeOf(base_schema, true)) {
+      continue;
+    }
+
+    Argument self_arg = schema->arguments()[0];
+    if (*self_arg.type() != *base_schema.arguments()[0].type()) {
+      continue;
+    }
```
Comment (eellison) [inline torch/csrc/jit/runtime/symbolic_shape_registry.cpp:662]: this should already be covered by the subtyping check

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 9]
```diff
@@ -629,6 +642,85 @@ std::unordered_map<const FunctionSchema*, std::shared_ptr<Graph>>
...
+    if (!self_arg.alias_info()->isWrite()) {
+      continue;
+    }
+    // Lastly check the return types are the same
+    if (schema->returns().size() != 1) {
+      continue;
+    }
```
Comment (eellison) [inline torch/csrc/jit/runtime/symbolic_shape_registry.cpp:669]: same with this

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 10]
```diff
@@ -629,6 +642,85 @@ std::unordered_map<const FunctionSchema*, std::shared_ptr<Graph>>
...
+    Argument ret_arg = schema->returns()[0];
+    if (!ret_arg.alias_info()->isWrite()) {
+      continue;
+    }
```
Comment (eellison) [inline torch/csrc/jit/runtime/symbolic_shape_registry.cpp:673]: ugh, i dont know if we really need to check this, it should be pretty alrready covered by firrst check but maybe it's finee

**Review by eellison — APPROVED** [review]
nice ! lgtm

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp (hunk 8 above)
Comment (Gamrix) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:662]: kk, will use the non-method version of the subtype check
